hbase-server/src/main/java/org/apache/hadoop/hbase/master/RegionsRecoveryConfigManager.java

```diff
@@ -0,0 +1,100 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.master;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.ChoreService;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.conf.ConfigurationObserver;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Config manager for RegionsRecovery Chore - Dynamically reload config and update chore
+ * accordingly
+ */
+@InterfaceAudience.Private
+public class RegionsRecoveryConfigManager implements ConfigurationObserver {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RegionsRecoveryConfigManager.class);
+
+  private final HMaster hMaster;
+  private int oldMaxStoreFileRefCount;
+  private int oldRegionsRecoveryInterval;
+
+  RegionsRecoveryConfigManager(final HMaster hMaster,
+      final Configuration conf) {
+    this.hMaster = hMaster;
+    this.oldMaxStoreFileRefCount = getMaxStoreFileRefCount(conf);
+    this.oldRegionsRecoveryInterval = getRegionsRecoveryChoreInterval(conf);
+  }
+
+  @Override
+  public void onConfigurationChange(Configuration conf) {
+    synchronized (this.hMaster) {
+      final int newMaxStoreFileRefCount = getMaxStoreFileRefCount(conf);
+      final int newRegionsRecoveryInterval = getRegionsRecoveryChoreInterval(conf);
+
+      LOG.info("Config Reload for RegionsRecovery Chore. oldMaxStoreFileRefCount: {}," +
+          " newMaxStoreFileRefCount: {}, oldRegionsRecoveryInterval: {}, " +
+          "newRegionsRecoveryInterval: {}", oldMaxStoreFileRefCount, newMaxStoreFileRefCount,
+        oldRegionsRecoveryInterval, newRegionsRecoveryInterval);
+
+      if (oldMaxStoreFileRefCount == newMaxStoreFileRefCount
+          && oldRegionsRecoveryInterval == newRegionsRecoveryInterval) {
+        // no need to re-schedule the chore with updated config
+        // as there is no change in desired configs
+        return;
+      }
+
+      RegionsRecoveryChore regionsRecoveryChore = new RegionsRecoveryChore(this.hMaster,
+        conf, this.hMaster);
+      ChoreService choreService = this.hMaster.getChoreService();
+
+      // Regions Reopen based on very high storeFileRefCount is considered enabled
+      // only if hbase.regions.recovery.store.file.ref.count has value > 0
+
+      if (newMaxStoreFileRefCount > 0) {
+        if (choreService.isChoreScheduled(regionsRecoveryChore)) {
```

Comment:
Isn't it easier to 'update' storeFileRefCountThreshold of an existing chore rather than creating a new one and canceling the running one? You'll need lesser synchronization (get rid of grabbing the HMaster monitor). All you probably need to do is make storeFileRefCountThreshold an AtomicInteger and add a helper there. Of course if nothing is running, you need to schedule a new one.

Reply:
Chore interval might also get updated which we might want to change too and also cancelChore() with 2nd arg false will ensure that if the chore was currently running then it is first completed and then unscheduled.

Reply:
nit: cancelChore() already checks for isChoreScheduled, so that can be omitted I guess.